### ISSUE TITLE
The efficiency of KNN

### ISSUE THREAD

**Issue by Ghostish** [QingyongHu/RandLA-Net#2]
Hi, I'm very interested in your work but I had a question about the computation of KNN.
The network is efficient due to the random sampling strategy. But I notice that the network still uses KNN to find each point's neighbor.  When the point cloud contains 10^6 points, calculating per point's KNN is very time-consuming.  Do you pre-calculate the KNN results before feeding point clouds to the network?

**Comment by Yang7879** [comment]
Hi @Ghostish, thanks for your interest in our paper.  Actually we use 10^5 pts in our experiments (clarified in  #1 and will be updated in the next version). 

We do not pre-compute the KNN. Instead, it's calculated within the neural network and the reported time in the paper (Sec 4.2) includes all operations such as neural processing, KNN and sampling of our framework.  Indeed, KNN is also expensive and it can take nearly half of the total time. We believe this could be the future work to further improve the overall efficiency.

**Comment by last-one** [comment]
Hi @Yang7879 , thanks for your great idea. But I also have the same question about KNN.

In Sec 4.2, the paper said the RandLA-Net takes the shortest time (translating to 23 frames per second) to infer the semantic labels for each large-scale point cloud (up to 10^6 points). Could you share the idea on computing the KNN. I don't find a way to compute the KNN of 10^6 points in 23 frames / s. Or do you compute the KNN of each frame in parallel? Could you share the inference time of each frame?

And looking forward to open source.

**Comment by Yang7879** [comment]
Hi @last-one , to compute KNN,  we use existing methods of NanoFLANN which relies on KD-trees and is therefore quite fast.  To infer the semantics of a large-scale pc in the SemanticKITTI dataset, we sample 10^5 points as a set and then feed as many sets as possible into the GPU to parallelly process. This allows our RandLA-Net to achieve 23 frames/s on average.  The code will be released soon.

For the implementation of KNN, you may refer: https://github.com/aboulch/ConvPoint

**Comment by last-one** [comment]
@Yang7879 Thanks for your reply. So could share the inference time of a single frame? Because in most cases, the time of parallel computing does not have much reference value.

**Comment by Yang7879** [comment]
@last-one There are many ways to evaluate the efficiency of different approaches. In our paper, for the fairest comparison that can be done, we obtain the total computation time for each approach, by treating it as a black box with the same validation set as the input and with the full utilization of the same hardware.  Therefore, the averaged 23 frames per second is a rough but reasonable indication for our efficient RandLA-Net. 

As to a strict frame rate you are eager to know, we are unfortunately unable to provide. It is non-trivial to conduct a strict experiment for all approaches, as it involves critical engineering tricks. For example, to what extent to parallelly process the point cloud? should we use single-float or double-float at inference? just to name a few. 

If you need a strict time for your engineering work,  you are encouraged to run experiments according to your specific hardware and software environments.

**Comment by tchaton** [comment]
@Yang7879, have a look at https://github.com/nicolas-chaulet/deeppointcloud-benchmarks
